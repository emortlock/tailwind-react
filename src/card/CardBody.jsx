import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

const CardBody = ({ children, className, ...rest }) => (
  <div {...rest} className={classnames('p-4', className)}>
    {children}
  </div>
)

CardBody.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
}

CardBody.defaultProps = {
  children: undefined,
  className: undefined,
}

export default CardBody
