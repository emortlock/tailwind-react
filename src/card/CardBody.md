For main content within a [`<Card />`](#card) component
