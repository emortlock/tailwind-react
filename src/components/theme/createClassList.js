import merge from 'lodash.merge'
import defaultTheme from './defaultTheme'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectTheme = {}) => {
  const theme = merge(defaultTheme, projectTheme)

  return [
    theme.radius,
    ...reduceMultiplePrefixes(theme.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(theme.container, 'max-w'),
    ...mapArray(theme.text.size.body, 'text'),
    ...mapArray(theme.text.size.title, 'text'),
    ...mapObject(theme.text.family, 'font'),
    ...reduceMultiplePrefixes(theme.brandColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(theme.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
