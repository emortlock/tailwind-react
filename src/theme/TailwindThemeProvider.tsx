import { Theme } from './TailwindTheme'

import React from 'react'
import merge from 'lodash.merge'

import { defaultTheme } from './defaultTheme'
import { TailwindTheme } from './TailwindTheme'

export const TailwindThemeProvider = ({
  theme = {},
  children,
}: {
  theme: Partial<Theme>
  children: React.ReactNode
}) => {
  const mergedTheme: Theme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}
