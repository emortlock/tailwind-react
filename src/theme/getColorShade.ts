const shades = ['100', '200', '300', '400', '500', '600', '700', '800', '900']

export const getColorShade = (
  color?: string | string[],
  offset: string | number = 1,
): string => {
  if (!color) return ''

  let currentColor: string | string[] = Array.isArray(color) ? color[0] : color

  if (offset === 0) return currentColor

  currentColor =
    currentColor === 'white' ? ['gray', '100'] : currentColor.split('-')
  let shadeOffset = offset

  if (color === 'white') {
    if (shadeOffset < 1) return color
    if (shadeOffset === 1) return 'gray-100'

    currentColor = ['gray', '100']
    if (typeof offset === 'number') shadeOffset = offset - 1
  }

  if (currentColor.length === 1) {
    currentColor.push('400')
  }

  if (typeof shadeOffset === 'string') {
    return `${currentColor[0]}-${shadeOffset}`
  }

  const shadeIndex = Math.min(
    Math.max(shades.indexOf(currentColor[1]) + shadeOffset, 0),
    shades.length - 1,
  )
  const newShade = shades[shadeIndex]

  return `${currentColor[0]}-${newShade}`
}
