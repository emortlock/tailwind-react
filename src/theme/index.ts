export * from './defaultTheme'
export * from './TailwindTheme'
export * from './TailwindThemeProvider'
export * from './withTheme'
export * from './useThemeValue'
export * from './getColorShade'
