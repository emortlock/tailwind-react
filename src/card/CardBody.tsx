import type { BoxProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'

import React from 'react'

import { RawBox as Box } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export type CardBodyProps<E extends HTMLElement = HTMLDivElement> = WithTheme &
  BoxProps<E>

export const RawCardBody = <E extends HTMLElement = HTMLDivElement>({
  theme,
  as = 'div',
  children,
  ...rest
}: CardBodyProps<E>) => (
  <Box<E> as={as} p={theme.spacing.md} {...rest}>
    {children}
  </Box>
)

export const CardBody = withTheme<CardBodyProps>(RawCardBody)
