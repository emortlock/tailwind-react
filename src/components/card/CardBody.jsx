import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Base } from '../primitives'

const CardBody = ({ theme, is, children, ...rest }) => (
  <Base is={is} p={theme.spacing.md} {...rest}>
    {children}
  </Base>
)

CardBody.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
}

CardBody.defaultProps = {
  is: 'div',
  children: undefined,
}

export { CardBody as component }
export default withTheme(CardBody)
