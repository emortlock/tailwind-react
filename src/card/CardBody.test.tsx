import React from 'react'
import { render, RenderResult } from '@testing-library/react'

import { RawCardBody as CardBody } from './CardBody'
import { defaultTheme } from '../theme'

const setup = (
  testProps = {},
): [RenderResult, { props: Record<string, any> }] => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = render(<CardBody {...props}>Hello World!</CardBody>)

  return [
    wrapper,
    {
      props,
    },
  ]
}

describe('CardBody', () => {
  it('renders matching snapshot', () => {
    const [{ asFragment }] = setup()

    expect(asFragment()).toMatchSnapshot()
  })
})
