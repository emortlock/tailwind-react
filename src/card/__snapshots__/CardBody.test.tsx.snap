// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardBody renders matching snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-4"
  >
    Hello World!
  </div>
</DocumentFragment>
`;
